### ISSUE TITLE
Some questions about the mask

### ISSUE THREAD

**Issue by Unispac** [salesforce/densecap#18]
Hello.
When I am reading your paper, I am a little bit confused about the following paragraph :

"Note that during decoding, the encoder performs the forward propagation again so that the representation of each encoder layer contains only the information for the current proposal. This is different from simply multiplying the mask with the existing representation from the encoder during proposal prediction, since the representation of the latter still contains information that is outside the proposal region."

It seems that you want to avoid including any information from outside of the proposal region.
But actually, there may be some corelations among events. I find that many previous work actually did take use of the context information from other events when captioning a particular one. To do so, they even use a Bi-LSTM...

Is there any other thought or trick about this?

**Comment by LuoweiZhou** [comment]
Hi @Unispac, this is a good question and thank you for the suggestion. We did study on the impact of having a single mask or a consistent [mask](https://github.com/salesforce/densecap/blob/master/model/transformer.py#L183) throughout the encoder and in terms of caption generation performance, the latter is better. We hypothesize that since self-attention is capable of attending on the entire input (usually long videos), the context might be even longer than our target proposal in terms of duration and brings in excessive info. In the meantime, LSTM/Bi-LSTM mainly attends on local context (e.g., see Fig. 7 in [Singh et al.](http://www.cs.umd.edu/~bharat/cvpr2016.pdf)) and is expected to not suffer from this issue.

**Comment by Unispac** [comment]
Thank you. @LuoweiZhou 
I guess I get your point.
You really did an excellent job!!

**Comment by LuoweiZhou** [comment]
@Unispac Thank you for the nice words! Please free feel to let us know if you have any other questions.

**Comment by Unispac** [comment]
@LuoweiZhou 
Hello, It's me again....
I have already carefully read your code.
But there are two parts which makes me feel weird...

Firstly, It's not difficult to understand that when performing training algorithm, you directly feed the GT Sentences  to the decoder and makes the GT words as the context to motivate the Decoder to generate the next word. But I notice that, when you use the GT words as the input for the decoder, you directly use all the T words once for all. It means when you generate the i th word, the i+1, i+2, ..... T th word will also be included as the context due to the self-attention mechanism... But actually, when we are in inference stage, we will generate the words one by one, and only use the information ahead of the i th word. 

Secondly, when calculating the cross-entropy loss between predicted_sentence and GT_sentence, you directly perform this : " sent_loss = F.cross_entropy(pred_sentence, gt_sent) * args.sent_weight ". But I found that, "pred_sentence" is the output from the function "forward" in "RealTransformer" and actaully the "pred_sentence" is a logits vector. Shouldn't we use the softmax probability to compute the cross-entropy with the GT words? 

I really feel lost ...

**Comment by LuoweiZhou** [comment]
@Unispac Please see below.
i) In the Transformer decoder, an upper triangular mask is applied to self-attention weights to make sure we do not look into future words when generating the current word (the autoregressive property). You can refer to [code](https://github.com/salesforce/densecap/blob/master/model/transformer.py#L103) and Sec. 3.2.3 in [Vaswani et al.](https://arxiv.org/pdf/1706.03762.pdf)
ii) [F.cross_entropy](https://pytorch.org/docs/stable/nn.functional.html#torch.nn.functional.cross_entropy) combines log_softmax and nll_loss.

**Comment by Unispac** [comment]
@LuoweiZhou 
Oh, I see. Thanks!
But I notice that the encoder is still limited to look into the future.
In general, I think it won't hurt if we conbine the future visual content for encoding...
Is this also based on the same consideration as that in proposal mask ?

**Comment by LuoweiZhou** [comment]
This is what we referred to as "a consistent mask throughout the (visual) encoder" in the caption generator here: https://github.com/salesforce/densecap/issues/18#issuecomment-520917586

**Comment by Unispac** [comment]
@LuoweiZhou 
Yeah, I get it.
By the way, when I am running the code, NAN appears in the loss during training..
How should I solve it.
